src/common/data/fields.d.ts

```diff
@@ -1 +1,372 @@
+  default: null;
+  nullable: false;
+  validate: typeof _validateId;
+  validationError: 'Invalid value provided for {name} id';
+}
+
+/**
+ * The standard identifier for a Document.
+ */
+export declare const DOCUMENT_ID: DocumentId;
+
+interface DocumentPermissions extends DocumentField<Partial<Record<string, ValueOf<typeof CONST.ENTITY_PERMISSIONS>>>> {
+  type: typeof Object;
+  required: true;
+  nullable: false;
+  default: { default: typeof CONST.ENTITY_PERMISSIONS.NONE };
+  validate: typeof _validatePermissions;
+  validationError: 'Invalid data structure encountered for {name} permissions';
+}
+
+/**
+ * The standard permissions object which may be included by a Document.
+ */
+export declare const DOCUMENT_PERMISSIONS: DocumentPermissions;
+
+/**
+ * Validate the structure of the permissions object: all anykeys are valid IDs and all values are permission levels
+ * @param perms - The provided permissions object
+ * @returns Is the object valid?
+ */
+declare function _validatePermissions(perms: object): boolean;
+
+interface ForeignDocumentField<
+  T extends {
+    type: {
+      readonly documentName: string;
+    };
+    required?: boolean;
+    nullable?: boolean;
+    default?: any;
+  }
+> extends DocumentField<string | null> {
+  type: typeof String;
+  required: T extends {
+    required: true;
+  }
+    ? true
+    : false;
+  nullable: T extends {
+    nullable?: true;
+  }
+    ? true
+    : T extends {
+        nullable: false;
+      }
+    ? false
+    : boolean;
+  default: T extends {
+    default: infer U;
+  }
+    ? U
+    : null;
+  clean: (d: unknown) => string | null;
+  validate: typeof _validateId;
+  validationError: `Invalid {name} {field} which must be a ${T['type']['documentName']} document id`;
+}
+
+/**
+ * Create a foreign key field which references a primary Document id
+ */
+export declare function foreignDocumentField<
+  T extends { type: { readonly documentName: string }; required?: boolean; nullable?: boolean; default?: any }
+>(options: T): ForeignDocumentField<T>;
+
+// TODO: Improve
+interface EmbeddedCollectionField<
+  ConcreteDocumentConstructor extends ConstructorOf<Document<any, any>>,
+  Options extends { required?: boolean; default?: any[] } = {}
+> extends DocumentField<any> {
+  type: Partial<Record<string, ConcreteDocumentConstructor>>;
+  required: Options extends { required?: true } ? true : Options extends { required: false } ? false : boolean;
+  default: Options extends { default?: Array<infer U> } ? Array<U> : never;
+  isCollection: true;
+}
+
+/**
+ * Create a special field which contains a Collection of embedded Documents
+ * @param document - The Document class definition
+ * @param options  - Additional field options
+ *                   (default: `{}`)
+ */
+export declare function embeddedCollectionField<
+  ConcreteDocumentConstructor extends { readonly documentName: string } & ConstructorOf<Document<any, any>>,
+  Options extends { required?: boolean; default?: any[] }
```

Comment:
The generic constraint here should also be shared between the interface and the function.

Reply:
I added another interface for the options here.